Add ES6 collection support to include()

shvaikalesh opened the pull request:
Closes #970

shvaikalesh commented on lib/chai/core/assertions.js:
```diff
@@ -504,10 +499,42 @@ module.exports = function (chai, _) {
+        break;
+
+      default:
+        var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
```
I feel like we should use `SameValue` here, but it is minor breaking change.

keithamus replied:
Well this is a new feature right? So no one would have hit this path before. It would be a small gotcha to say:

> If you pass in (weak)?(map|set) then it'll use samevalue, otherwise ===

keithamus replied:
Yes - we should look to change many of these instances for chai@5 i think

shvaikalesh commented on lib/chai/core/assertions.js:
```diff
@@ -504,10 +499,42 @@ module.exports = function (chai, _) {
+
+      case 'weakmap':
+      case 'weakset':
+        if (isDeep) {
```
I think we should consider containing code like this into a private method: it is easy to forget to put `flagMsg`, pass `ssfi`, or skip second argument.

keithamus replied:
Agreed, but I'd like to see focus on #585 to get the assertion functions to be pure functions returning objects, so we can wrap them up in assertionerrors

keithamus commented:
I'm very happy with this, but let's get another review to sign off before merge. @meeber @lucasfcosta?

vieiralucas commented on lib/chai/core/assertions.js:
```diff
@@ -504,10 +499,42 @@ module.exports = function (chai, _) {
+        var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
+        obj.forEach(function (item) {
+          included = included || isEql(item, val);
+        });
```
Can we do something here to optimize this by breaking out the loop once `included` is true?

I realize that then we would need to have 3 implementations, one for `Array`, another for `Map` and another for `Set`.

Do you think it is worth it?

shvaikalesh replied:
We can optimize away looping through `map` and `set` for non-deep `include` (cool) and use `some` + `indexOf` on arrays. Code:

```diff
+  case 'map':
+  case 'set':
+    if (isDeep) {
+      obj.forEach(function (item) {
+        included = included || _.eql(item, val);
+      });
+   } else {
+     included = obj.has(val);
+   }
+   break;
+
+  case 'array':
+    if (isDeep) {
+     included = obj.some(function (item) {
+       return _.eql(item, val);
+     })
+    } else {
+      included = obj.indexOf(val) !== -1;
+    }
+    break;

-  default:
-    var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
-    obj.forEach(function (item) {
-      included = included || isEql(item, val);
-    });
```

keithamus replied:
This looks good but I'd be a little careful here with IE11 - which has terrible support for Map/Set. It probably will work but I've often been surprised how seemingly obvious features are missing from IE11's Map/Set

shvaikalesh commented:
One more thing: should we throw when `deep.include` is called on string? Currently we just ignore the flag, but the assertion does not make sense.

keithamus commented:
@shvaikalesh I think we should - as a general future feature consideration - start adding warnings for wonky assertion chains, rather than explicitly erroring. Let's leave it out of this PR and discuss it in https://github.com/chaijs/chai/issues/996

shvaikalesh commented:
I've updated the docs and dropped `WeakMap` support -- we can't iterate over it's values to make an assertion.

Regarding `deep` flag: point raised by @meeber in https://github.com/chaijs/chai/issues/996#issuecomment-310481585 is valid for `WeakSet` targets too. I am fine with silently ignoring it.

lucasfcosta commented on lib/chai/core/assertions.js:
```diff
@@ -504,10 +513,62 @@ module.exports = function (chai, _) {
+
+      case 'map':
+        var isEql = isDeep ? _.eql : SameValueZero;
+        obj.forEach(function (item) {
```
As @vieiralucas said, we might be able to optimize this by getting out of the loop when we found the included item.

In order to support IE and older versions of it we can `throw` an error inside the `forEach` callback and catch it.

This is a cross-platform compatible implementation I've made on `Sinon` for the `every` method. We can inspire our `some` implementation in that. However, even though I'd like it to be done, I won't block merging this PR because of that.

meeber replied:
@lucasfcosta I think this is worth exploring. I'd prefer this kind of optimization technique be implemented in a separate PR as a refactor.

lucasfcosta replied:
@meeber agreed.

LGTM! Let's get this merged.

lucasfcosta commented on lib/chai/core/assertions.js:
```diff
@@ -330,6 +330,16 @@ module.exports = function (chai, _) {
+   *     expect(new Set([1, 2])).to.include(2);
+   *
+   * When the target is a Map, `.include` asserts that the given `val` is one of
+   * the values of the target. SameValueZero equality algorithm is used.
```
I think it would be good to explain how the `SameValueZero` algorithm works. I think many people could get confused by that and would end up either having to read the code to understand it or just avoiding using this assertion because they're not sure of what it does.

meeber replied:
@lucasfcosta Fair point. In this case, I would suggest moving the phrase "SameValueZero equality algorithm is used" from where it is currently down to the separate paragraph where `strict` and `deep` equality` is discussed. This would make the doc style more consistent anyway. And then rather than explain what "SameValueZero equality" is, we could provide a link to an explanation, just like we do with the "deep equal" algorithm.

keithamus replied:
We could just refer to is as it is more commonly known; [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) or if you'd like to we coud keep the phrasing but link to [the equality article on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness) which mentions `SameValueZero`.
